libmscore/scale.cpp

```diff
@@ -0,0 +1,265 @@
+
+      fillMissing12();
+      }
+
+//---------------------------------------------------------
+//   fillMissing12
+//---------------------------------------------------------
+void Scale::fillMissing12() {
+      computedNotes[TPC_D_B] = computedNotes[TPC_C_S];
+      computedNotes[TPC_E_B] = computedNotes[TPC_D_S];
+      computedNotes[TPC_G_B] = computedNotes[TPC_F_S];
+      computedNotes[TPC_A_B] = computedNotes[TPC_G_S];
+      computedNotes[TPC_B_B] = computedNotes[TPC_A_S];
+
+      fillMissing17();
+      }
+
+//---------------------------------------------------------
+//   fillMissing17
+//---------------------------------------------------------
+void Scale::fillMissing17() {
+      computedNotes[TPC_E_S] = computedNotes[TPC_F_B] = (computedNotes[TPC_E] + computedNotes[TPC_F]) / 2;
+      computedNotes[TPC_B_S] = computedNotes[TPC_C_B] = (computedNotes[TPC_B] + computedNotes[TPC_C]) / 2;
+
+      fillMissing21();
+      }
+
+//---------------------------------------------------------
+//   fillMissing21
+//---------------------------------------------------------
+void Scale::fillMissing21() {
+      computedNotes[TPC_B_SS] = computedNotes[TPC_D_BB] = computedNotes[TPC_C];
+      computedNotes[TPC_C_SS] = computedNotes[TPC_E_BB] = computedNotes[TPC_D];
+      computedNotes[TPC_D_SS] = computedNotes[TPC_F_BB] = computedNotes[TPC_E];
+      computedNotes[TPC_E_SS] = computedNotes[TPC_G_BB] = computedNotes[TPC_F];
+      computedNotes[TPC_F_SS] = computedNotes[TPC_A_BB] = computedNotes[TPC_G];
+      computedNotes[TPC_G_SS] = computedNotes[TPC_B_BB] = computedNotes[TPC_A];
+      computedNotes[TPC_A_SS] = computedNotes[TPC_C_BB] = computedNotes[TPC_B];
+      }
+
+//---------------------------------------------------------
+//   getTuning
+//---------------------------------------------------------
+
+double Scale::getTuning(const Note* note) const
+      {
+      static map<Tpc, double> standardNotes = {
+            {TPC_D_BB, 1200}, {TPC_C_S, 100},  {TPC_D_B, 100},  {TPC_C_SS, 200},  {TPC_D, 200},
+            {TPC_E_BB, 200},  {TPC_D_S, 300},  {TPC_E_B, 300},  {TPC_D_SS, 400},  {TPC_E, 400},
+            {TPC_F_BB, 400},  {TPC_E_S, 500},  {TPC_F_B, 400},  {TPC_E_SS, 500},  {TPC_F, 500},
+            {TPC_G_BB, 500},  {TPC_F_S, 600},  {TPC_G_B, 600},  {TPC_F_SS, 700},  {TPC_G, 700},
+            {TPC_A_BB, 700},  {TPC_G_S, 800},  {TPC_A_B, 800},  {TPC_G_SS, 900},  {TPC_A, 900},
+            {TPC_B_BB, 900},  {TPC_A_S, 1000}, {TPC_B_B, 1000}, {TPC_A_SS, 1100}, {TPC_B, 1100},
+            {TPC_C_BB, 1100}, {TPC_B_S, 1200}, {TPC_C_B, 1100}, {TPC_B_SS, 1200}, {TPC_C, 1200}};
+
+      auto standardNote = standardNotes.find((Tpc)note->tpc1());
+      auto computedNote = computedNotes.find((Tpc)note->tpc1());
+
+      if (standardNote == standardNotes.end() || computedNote == computedNotes.end()) {
+            return 0;
+            }
+
+      return standardNote->second - computedNote->second;
+      }
+}
```

Comment:
Less code means less code to understand, less things which could go wrong and so on. That code proceeds by logical steps and it is not only more compact: it supports any (present or future) sequence of TPC, not only the 5 currently implemented, with the only requirement that the sequence shall not have 'holes'. It is necessary to understand it once for all, not each time a new sequence is added. If you don't like the 'big' formula to convert octave offset to 12EDO delta, see below.

I do not understand your comment about new symbols: new TPC's (for instance in the unlikely case MuseScore is going to support triple accidentals) would fall in the logic. As well as for new types of scales. In any case, it would simply be a matter of adding more elements to the `tpcs[][]` array and/or to the `supportedNbs` map at the top.

About your points:
- Correct: I left out a `* centsPerSemitone` factor while copying my notes; as I noted, this is not production code, but something I wrote mostly by memory to show the coding patterns. Note also that for the TPC_F_BB case, the expected value is 300, not 400 and that `tpc - TPC_MIN` equals 0, not 1. So, for this case the reckoning comes out as `((3 + 0*7) * 100) % 1200 = 300`, as expected.

As an aside: if a lookup table (`static const int equalTempCent[TPC_NUM_OF];`) filled with the 12 EDO cents in TPC order is preferred to this formula, ok as well: it would probably spare some CPU cycle. I made up the formula mainly to show that, by using TPC's, everything falls 'automagically' in its place and almost anything can be computed with general rules, rather than with _ad hoc_ cases.
- I **am** supporting the 17 note case: it is the third element of the array at the top (your code is filtering it out, while checking the number of notes at the top, though).
- Can you explain why TPC_F_BB, TPC_C_BB, TPC_E_SS, TPC_B_SS are not set correctly?
- No need to check double accidentals in particular: they fall in the general logic. Anything below `tpcMin` or above `tpcMax` is for sure not set yet.
- Correct: read `for (i = tpcMin-1; i <= TPC_E_B; i--)` instead. Of course this does not set TPC_F_B and TPC_C_B: they do not fall within the minimum 12-note range!
- Symmetrically to the above.